lib/aio/job.py

```diff
@@ -0,0 +1,189 @@
+
+        # reporting
+        self.context = get_str(obj, 'context', None)
+        self.slug = get_str(obj, 'slug', None)
+        self.title = get_str(obj, 'title', None)
+        self.report = get_dict(obj, 'report', None)
+
+
+async def timeout_minutes(minutes: float) -> Never:
+    await asyncio.sleep(60 * minutes)
+    raise Failure(f'Timeout after {minutes} minutes')
+
+
+async def poll_pr(api: Forge, repo: str, pull_nr: int, expected_sha: str) -> Never:
+    while True:
+        if reason := await api.check_pr_changed(repo, pull_nr, expected_sha):
+            raise Failure(reason)
+        await asyncio.sleep(60)
+
+
+async def run_container(job: Job, subject: Subject, ctx: JobContext, log: LogStreamer) -> None:
+    with tempfile.TemporaryDirectory() as tmpdir_path:
+        tmpdir = Path(tmpdir_path)
+        cidfile = tmpdir / 'cidfile'
+        attachments = tmpdir / 'attachments'
+
+        args = [
+            *ctx.container_cmd, 'run',
+            *ctx.container_run_args,
+            f'--cidfile={cidfile}',
+            *(f'--env={k}={v}' for k, v in job.env.items()),
+            '--env=TEST_ATTACHMENTS=/tmp/attachments',
+            f'--env=COCKPIT_CI_LOG_URL={log.url}',
+            *itertools.chain.from_iterable(args for name, args in ctx.secrets_args.items() if name in job.secrets),
+
+            job.container or ctx.default_image,
+
+            'python3', '-c', Path(f'{BOTS_DIR}/checkout-and-run').read_text(),  # lulz
+            f'--revision={subject.sha}'
+        ]
+        if subject.rebase:
+            args.append(f'--rebase={subject.rebase}')
+        args.append(f'{subject.clone_url}')
+
+        if job.command:
+            args.append('--')
+            args.extend(job.command)
+
+        async with spawn(args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.STDOUT) as container:
+            try:
+                # Log data until we hit EOF
+                assert container.stdout is not None
+                async for block in read_utf8(container.stdout):
+                    log.write(block)
+
+                # Upload all attachments
+                # TODO: live updates
+                # TODO: invent async tarfile for StreamReader
+                with contextlib.suppress(FileNotFoundError):
+                    cid = cidfile.read_text().strip()
+                    await run([*ctx.container_cmd, 'cp', '--', f'{cid}:/tmp/attachments/.', f'{attachments}'])
+                    for file in attachments.rglob('*'):
+                        log.index.write(str(file.relative_to(attachments)), file.read_bytes())
+
+                if returncode := await container.wait():
+                    raise Failure(f'Container exited with code {returncode}')
+
+            finally:
+                await run([*ctx.container_cmd, 'rm', '--force', '--time=0', f'--cidfile={tmpdir}/cidfile'],
+                          stdout=asyncio.subprocess.DEVNULL, stderr=asyncio.subprocess.DEVNULL)
+
```

Comment:
I think I added this redirection before removing `--rm` on the container.  At that time, it was meant as a mop-up-in-case-of-crash rather than the main way of killing the container, and would give the error message about the container missing in the usual case.  That's no longer true, so I'll fix it.  Thanks for the suggestion.

Reply:
In, fact, why didn't you just use `run --rm`? Isn't that easier and more robust?

Reply:
The container needs to survive the test run exiting in order for us to retrieve the attachments.

This is tricky, and something we talked about before.  Even if we move to a model where we support live attachments again, we face the possibility of an attachment being written at the end of a test run and the container immediately exiting.

That implies that we need to accept one of the following:
 - we need a streaming protocol that's part of the data that the main process of the container delivers before it exits, either via stdout/err or some sort of fd hacks; or
 - we need attachments stored in a volume that survives the container `rm`; or
 - we can't use `--rm`.

Reply:
Of all these, creating a `temp_dir/attachments` on the host and mounting that as a volume feels best to me. That part can be a follow-up.

Reply:
I actually like the idea of having attachments as something done more explicitly via testlib API, with strong assurances that everything is dealt with by the time the main process exits (ie: no longer based on having a directory at all, and not involving the possibility of having stuff hanging around after the container exits).  This starts sounding like tar-via-passed-in-fd or so, but maybe we could find another way...

Reply:
Straw man:
 - `test/common/run-tests` checks if `TEST_ATTACHMENT_FD` is set, and if so, creates a temporary directory and sets the `TEST_ATTACHMENTS` envvar pointing to it
 - on exit, it calls `tar` on the contents of that directory directing output to the fd given in the envvar
 - from job-runner we use tarfile directly into s3streamer and nothing ever touches the disk outside of the container
 - we can still use the `TEST_ATTACHMENTS` directory approach in places where fd passing isn't supported

Definitely a follow-up, though.  I realized quite early in working on this that the millions of ways to handle attachments would be a time-sucking pit and quickly decided to slap the `# TODO` comment on it and move on!!

Reply:
Agreed to postpone this. That'll be a longer discussion, and I'll accept the non-live attachments regression for now.

Reply:
Ah, you already fixed that in the latest force-push.